### ISSUE TITLE
Injecting Data.Map

### ISSUE THREAD

**Issue by sjakobi** [dhall-lang/dhall-haskell#1392]
In https://github.com/dhall-lang/dhall-haskell/pull/1391 I'm adding an `Interpret` instance. It would be nice to have an `Inject` instance too.

Any `Data.Map` with "injectable" keys and values can be represented as `Prelude.Map.Type` list of `mapKey`-`mapValue` records. Eg.

```
fromList [ (1, False) ]

=>

[ { mapKey = 1, mapValue = False } ]
```

So I think that's what the `Inject` instance should do.

But it would also be nice if a `Data.Map Text v` could be embedded as  record. We should add an `InputType` for this, and possibly a newtype wrapper with the matching `Inject` instance. Maybe call it `RecordMap`?!

@Gabriel439 Thoughts?

**Comment by Gabriel439** [comment]
@sjakobi: I definitely agree that there should be an `Inject` instance for `Map` and I also like the idea of the  `RecordMap` `newtype` to get the equivalent record

**Comment by sjakobi** [comment]
Regarding the `RecordMap` wrapper: I wonder whether we should allow any `IsString` key type here, or whether it's better to constrain it to `Text` keys… The latter could be more efficient.

Or we could rely on something like [this `ToText` class](http://hackage.haskell.org/package/text-conversions-0.3.0/docs/Data-Text-Conversions.html#t:ToText).

**Comment by Gabriel439** [comment]
@sjakobi: I think constraining it to `Text` is fine for now.  We can always generalize it later if users need it

**Comment by jiegillet** [comment]
I'd like to try this one, after #1395 is done.

**Comment by jiegillet** [comment]
Is this cheating?

```haskell
data InputRecord k v =
    InputRecord { mapKey :: k, mapValue :: v } deriving Generic

instance (Inject k, Inject v) => Inject (InputRecord k v)

instance (Inject k, Inject v) => Inject (Map k v) where
    injectWith =
        let mapToInputRecords = fmap (uncurry InputRecord) . Data.Map.toList
        in fmap (contramap mapToInputRecords) injectWith
```

it works:

```
ghci> prettyExpr $ embed inject $ Data.Map.fromList [(1, False)]
[ { mapKey = +1, mapValue = False } ]
```

**Comment by sjakobi** [comment]
> Is this cheating?

No. :) I think it's borderline _too_ smart though. I also wonder whether these `Generic` instances are less efficient than manually written code…


> By the way, this project is really solid and a pleasure to learn.

I agree. @Gabriel439 is a great maintainer and open source shepherd! :+1:

**Comment by Gabriel439** [comment]
@sjakobi: Thank you! 🙂

**Comment by jiegillet** [comment]
I need a hand on the `RecordMap` instance. 
In the `InputType embed declared`, the `embed` part is no problem, but I don't think the `declared` is possible, since the fields of a record are part of its type, and if the fields come from any given `Map k v`,  the `k` fields cannot be extracted in general and so there cannot be a type.
Am I missing something?

**Comment by sjakobi** [comment]
Oh, right. That `RecordMap` idea doesn't work then.

We could consider generalizing `Inject.declared` to allow types that depend on the result of `embed`, but that would probably cause other problems.

**Comment by jiegillet** [comment]
OK, I'll scratch the idea for now.

**Comment by jiegillet** [comment]
Fixed in #1412
